Add release notes for April 17, 2026 (#4023)
514-labs/moose: apps/framework-docs-v2/content/moosestack/release-notes/2026-04-17.mdx

---
title: April 17, 2026
description: Release notes for April 17, 2026
---

import { Callout } from "@/components/mdx";

# April 17, 2026

## 🐻 Boreal

### New Features

> **Add build logs support to 514 logs command** ([@LucioFranco](https://github.com/LucioFranco))
>   Added a new RPC handler that enables users to fetch and view build logs through the 514 logs command, providing better visibility into deployment build processes and debugging capabilities.
> 

Reviewer: ```suggestion
```

          client.command(f"CREATE OR REPLACE VIEW {view_name} AS {sql_definition}")
          print(f"✓ Created view: {view_name}")
      
      print("All analytics views ready!")
  ```

  This example shows how to use lifecycle hooks to automatically set up SQL views after Moose infrastructure is deployed. The setup script runs on first start and after every reload, ensuring views are always up-to-date during development.

  PR: [#4008](https://github.com/514-labs/moose/pull/4008) | Docs: [/moosestack/configuration/dev-environment](https://docs.moosejs.com/moosestack/configuration/dev-environment)

**Add ClickHouse Dictionary support (core infrastructure)** ([@514Ben](https://github.com/514Ben))
  Adds foundational support for ClickHouse dictionaries in the Moose framework, including dictionary creation, modification, and migration capabilities. This enables users to define and manage dictionaries alongside their data models, though the user-facing API is still in development (this is part 1 of 6 PRs).

  PR: [#3997](https://github.com/514-labs/moose/pull/3997) | Docs: [/moosestack/olap](https://docs.moosejs.com/moosestack/olap)

> **Enhanced migration failure diagnostics with drift detection** ([@onelesd](https://github.com/onelesd))
>   Migration failures now provide detailed forensics showing which database operations succeeded, failed, and were skipped, along with automatic drift detection when the database state diverges from migration expectations. This helps developers understand exactly what happened during partial migration failures and provides clear recovery guidance, especially important since ClickHouse DDL operations cannot be rolled back transactionally.
> 

Reviewer 2: ```suggestion
```

this is part of a larger feature that shouldn't be mentioned in part like this

  PR: [#3999](https://github.com/514-labs/moose/pull/3999) | Docs: [/moosestack/data-modeling](https://docs.moosejs.com/moosestack/data-modeling)

> **Add embedded Kafka broker for development** ([@LucioFranco](https://github.com/LucioFranco))
>   Introduced devkafka, an embedded Kafka broker that runs locally during development. This eliminates the need to install and manage a separate Kafka instance when developing streaming applications with Moose. The embedded broker supports core Kafka operations including topic management, message production/consumption, and consumer group coordination, making it easier to test and develop data pipelines locally.
> 
>   PR: [#3933](https://github.com/514-labs/moose/pull/3933) | Docs: [/moosestack/data-modeling](https://docs.moosejs.com/moosestack/data-modeling)
> 
> **Add embedded Redis server for local development** ([@LucioFranco](https://github.com/LucioFranco))
>   Adds devredis, an embedded Redis server that runs automatically during local development. This eliminates the need for developers to install and manage Redis separately when working on MooseStack applications locally. The embedded server supports core Redis operations including data storage, lists, pub/sub, and Lua scripting, providing a seamless development experience without external dependencies.
> 
>   PR: [#3932](https://github.com/514-labs/moose/pull/3932) | Docs: [/moosestack/moose-cli](https://docs.moosejs.com/moosestack/moose-cli)
> 

Reviewer: ```suggestion
```

> **Support table constraints in data models** ([@phiSgr](https://github.com/phiSgr))

Reviewer 3: original work was from
https://github.com/514-labs/moosestack/pull/3810
not me

  **Monorepo Docker configuration with context_path**

  ```toml
  # moose.config.toml - Analytics service in a monorepo
  project_name = "analytics"
  
  [docker_config]
  # Enable custom Dockerfile for monorepo build optimizations
  custom_dockerfile = true
  dockerfile_path = "./Dockerfile"
  # Set build context to monorepo root to access shared packages
  context_path = "../../.."
  
  [infrastructure]
  # Analytics pipeline infrastructure
  olap_database = "clickhouse"
  streaming_engine = "redpanda"
  ```

  Shows how to configure a Moose project in a monorepo using context_path to access shared packages and workspace files from a custom Dockerfile. The context_path setting allows the Docker build to reference files outside the project directory.

  PR: [#3947](https://github.com/514-labs/moose/pull/3947) | Docs: [/moosestack/configuration/docker](https://docs.moosejs.com/moosestack/configuration/docker)

> **Add native infrastructure provider with --dockerless flag** ([@LucioFranco](https://github.com/LucioFranco))

Reviewer: I would almost say promote this up?

> **Add native infrastructure provider with --dockerless flag** ([@LucioFranco](https://github.com/LucioFranco))

Reviewer: ```suggestion
**Add dockerless provider for `moose dev` with the new alpha --dockerless flag** ([@LucioFranco](https://github.com/LucioFranco))
```

>   Introduces a new --dockerless flag for the dev command that allows running ClickHouse and Temporal as native binaries instead of Docker containers. This provides an alternative for developers who prefer not to use Docker or encounter Docker-related issues. The feature automatically downloads and manages the required binaries (ClickHouse v25.8.18.1-lts and Temporal CLI v1.3.0) in a local cache, making local development more accessible and reducing Docker dependencies.

Reviewer: ```suggestion
  Introduces a new --dockerless flag for the dev command that allows running all of the dependencies for`moose dev` without the need for docker.
```
